pkg/dataobj/internal/dataset/read_stats.go

```diff
@@ -0,0 +1,288 @@
+	ds.SecondaryColumnUncompressedBytes = 0
+}
+
+// ReaderStats tracks statistics about dataset read operations.
+type ReaderStats struct {
+	// TODO(ashwanth): global stats from [stats.Context] can be updated by the
+	// engine at the end of the query execution once we have a stats collection
+	// framework integrated with the execution pipeline.
+	// Below reference is a temporary stop gap.
+	//
+	// Reference to [stats.Context] to update relevant global statistics.
+	globalStats *stats.Context
+
+	// Column statistics
+	PrimaryColumns   uint64 // Number of primary columns to read from the dataset
+	SecondaryColumns uint64 // Number of secondary columns to read from the dataset
+
+	// Page statistics
+	PrimaryColumnPages   uint64        // Total pages in primary columns
+	SecondaryColumnPages uint64        // Total pages in secondary columns
+	DownloadStats        DownloadStats // Download statistics for primary and secondary columns
+
+	ReadCalls int64 // Total number of read calls made to the reader
+
+	// Row statistics
+	MaxRows                uint64 // Maximum number of rows across all columns
+	RowsToReadAfterPruning uint64 // Total number of primary rows to read after page pruning
+
+	PrimaryRowsRead   uint64 // Actual number of primary rows read.
+	SecondaryRowsRead uint64 // Actual number of secondary rows read.
+
+	PrimaryRowBytes   uint64 // Total bytes read for primary rows
+	SecondaryRowBytes uint64 // Total bytes read for secondary rows
+}
+
+type ctxKeyType string
+
+const (
+	readerStatsKey ctxKeyType = "reader_stats"
+)
+
+func WithStats(ctx context.Context, stats *ReaderStats) context.Context {
+	if stats == nil {
+		return ctx
+	}
+
+	return context.WithValue(ctx, readerStatsKey, stats)
+}
+
+// StatsFromContext returns the reader stats from context.
+func StatsFromContext(ctx context.Context) *ReaderStats {
+	v, ok := ctx.Value(readerStatsKey).(*ReaderStats)
+	if !ok {
+		return &ReaderStats{}
+	}
+	return v
+}
+
+// IsStatsPresent returns true if reader stats are present in the context.
+func IsStatsPresent(ctx context.Context) bool {
+	_, ok := ctx.Value(readerStatsKey).(*ReaderStats)
+	return ok
+}
+
+func (s *ReaderStats) LinkGlobalStats(stats *stats.Context) {
+	// If the global stats are already set, we don't override it.
+	if s.globalStats == nil {
+		s.globalStats = stats
+	}
+}
+
+func (s *ReaderStats) AddReadCalls(count int) {
+	s.ReadCalls += int64(count)
+}
+
+func (s *ReaderStats) AddPrimaryColumns(count uint64) {
+	s.PrimaryColumns += count
+}
+
+func (s *ReaderStats) AddSecondaryColumns(count uint64) {
+	s.SecondaryColumns += count
+}
+
+func (s *ReaderStats) AddPrimaryColumnPages(count uint64) {
+	s.PrimaryColumnPages += count
+}
+
+func (s *ReaderStats) AddSecondaryColumnPages(count uint64) {
+	s.SecondaryColumnPages += count
+}
+
+func (s *ReaderStats) AddPrimaryRowsRead(count uint64) {
+	s.PrimaryRowsRead += count
+	if s.globalStats != nil {
+		s.globalStats.AddPrePredicateDecompressedRows(int64(count))
```

Comment:
By the way, feel free to rename or remove these global stat fields if they don't make sense anymore. We don't rely on them anywhere and they are for our own use. You can even just drop them from the proto as we don't store them beyond the lifetime of a query